### PR TITLE
Add build scripts

### PR DESCRIPTION
**Pull request by dbatyai**
I've added some scripts for building and running tests, and in the meantime I also improved option handling in cmake. Please take a look. Suggestions are welcome.

Issue: #6

### CODE REVIEW

File: tools/scripts/build.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    return opts
+
+
+# Perform the build
+def build(arguments):
+    basedir = path.join(getcwd(), path.split(sys.argv[0])[0], '..', '..')
+    build_path = path.join(basedir, 'build', arguments.build_type)
+
+    call(['mkdir', '-p', build_path])
```
Comment (galpeter) [inline tools/scripts/build.py]: really? calling out to shell to create dirs? why not: https://docs.python.org/2/library/os.html#os.makedirs

Comment (dbatyai) [inline tools/scripts/build.py]: Good point

File: tools/scripts/unittest.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+def run_unittest(type):
+    basedir = path.join(getcwd(), path.split(sys.argv[0])[0], '..', '..')
+    build_path = path.join(basedir, 'build', type)
+    if not path.isfile(path.join(build_path, 'bin', 'unit')):
+        print "Error: Unit-tests are not built."
+        return 1
+
+    chdir(build_path)
+    return call(['./bin/unit'])
```
Comment (szledan) [inline tools/scripts/unittest.py]: @kkristof : Is `'./bin/unit'` a platform independent form?

Comment (kkristof) [inline tools/scripts/unittest.py]: Not really, but the os.path is.

**Comment by kkristof** [comment]
Apart of the lack of the `os.path` in the `unittest.py:18` the patch looks good. Probably @galpeter could verify that.

File: tools/scripts/build.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+def build(arguments):
+    basedir = path.join(getcwd(), path.split(sys.argv[0])[0], '..', '..')
+    build_path = path.join(basedir, 'build', arguments.build_type)
+
+    try:
+        makedirs(build_path)
+    except OSError:
+        pass
+
+    chdir(build_path)
```
Comment (elecro) [inline tools/scripts/build.py]: Do we really need to change to the build path? can't we use the `make -C <workdir>` option?

File: tools/scripts/build.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    if ret:
+        return ret
+
+    return call(['make']);
+
+
+# Print build result
+def print_result(ret):
+    print('')
+    print('------------------------------')
```
Comment (elecro) [inline tools/scripts/build.py]: Could be: `print('-' * 31)`

File: tools/scripts/build.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    if ret:
+        print 'Build failed with exit code: ' + str(ret)
```
Comment (elecro) [inline tools/scripts/build.py]: Maybe add parentheses? and also we could use formatting: '... exit code: %s' % (ret)

File: tools/scripts/build.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    chdir(build_path)
+
+    if arguments.clean:
+        call(['make', 'clean'])
+
+    ret = call(['cmake'] + create_options(arguments) + [basedir])
+    if ret:
+        return ret
+
+    return call(['make']);
```
Comment (elecro) [inline tools/scripts/build.py]: No need for the `;`

File: tools/scripts/build.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+#! /usr/bin/python2
+
+import argparse
+import sys
+from os import path
+from os import chdir
+from os import getcwd
+from os import makedirs
+from subprocess import call
```
Comment (elecro) [inline tools/scripts/build.py]: IMHO we should just `import subprocess` and then use `subprocess.call`

File: tools/scripts/build.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
+#! /usr/bin/python2
```
Comment (elecro) [inline tools/scripts/build.py]: Is there any specific thing why this should be only python2?

File: tools/scripts/build.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    if (arguments.backend != 'gles2'):
+        opts.append('-DUSE_GLES2=OFF')  #GLES2 is enabled by default, so just disable it if we're not using it.
+
+    return opts
+
+
+# Perform the build
+def build(arguments):
+    basedir = path.join(getcwd(), path.split(sys.argv[0])[0], '..', '..')
```
Comment (elecro) [inline tools/scripts/build.py]: Here you want to get the base dir of the python script?

File: tools/scripts/unittest.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+from subprocess import call
+
+def run_unittest(type):
+    basedir = path.join(getcwd(), path.split(sys.argv[0])[0], '..', '..')
+    build_path = path.join(basedir, 'build', type)
+    if not path.isfile(path.join(build_path, 'bin', 'unit')):
+        print "Error: Unit-tests are not built."
+        return 1
+
+    chdir(build_path)
```
Comment (elecro) [inline tools/scripts/unittest.py]: Do we need to change the change the working dir?

File: tools/scripts/build.py (hunk 9 above)
Comment (dbatyai) [inline tools/scripts/build.py]: Basically, yes. It's not very pretty, if you know a better way, then please share.

Comment (elecro) [inline tools/scripts/build.py]: This could help: http://www.blog.pythonlibrary.org/2013/10/29/python-101-how-to-find-the-path-of-a-running-script/

**Comment by dbatyai** [comment]
I updated the patch based on the comments.

**Comment by kkristof** [comment]
Looks good to me.

File: .gitignore [hunk 11]
```diff
@@ -1,2 +1,3 @@
 build
-
+tools/scripts/*.pyc
+tools/scripts/__pycache__
```
Comment (elecro) [inline .gitignore]: IMHO just `*.pyc` and `__pycache__` would be better, no need for the full path

File: tools/scripts/build.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+    try:
+        makedirs(build_path)
+    except OSError:
+        pass
+
+    if arguments.clean:
+        subprocess.call(['make', '-C', build_path, 'clean'])
+
+    ret = subprocess.call(['cmake', '-E', 'chdir', build_path, 'cmake'] + create_options(arguments) + [basedir])
```
Comment (elecro) [inline tools/scripts/build.py]: chdir? hmm feels strange

Comment (elecro) [inline tools/scripts/build.py]: I think we can simply use: `cmake -H<source_path> -D<build_output_path> ...`

File: tools/scripts/cppcheck.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
+#! /usr/bin/python
+
+import subprocess
+import sys
+from os import path
+from os import chdir
```
Comment (elecro) [inline tools/scripts/cppcheck.py]: The chdir is not used at all in the file.

File: tools/scripts/build.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
+#! /usr/bin/python
+
+import argparse
+import subprocess
+import sys
+from os import path
+from os import chdir
```
Comment (elecro) [inline tools/scripts/build.py]: The chdir is not used at all.

File: tools/scripts/build.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+    if (arguments.backend != 'gles2'):
+        opts.append('-DUSE_GLES2=OFF')  #GLES2 is enabled by default, so just disable it if we're not using it.
+
+    return opts
+
+
+# Perform the build
+def build(arguments):
+    basedir = path.abspath(path.dirname(path.dirname(path.join(getcwd(), path.dirname(sys.argv[0])))))
```
Comment (elecro) [inline tools/scripts/build.py]: I would go with this one:
```python
basedir = path.abspath(path.join(path.dirname(__file__), "..", ".."))
```

File: tools/scripts/build.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+from os import getcwd
+from os import makedirs
+
+
+# Create cmake option list from parsed arguments
+def create_options(arguments):
+    opts = []
+    opts.append('-DCMAKE_BUILD_TYPE=' + arguments.build_type.capitalize())
+    opts.append('-DUSE_' + arguments.backend.upper() + '=ON')
+    opts.append('-DLOG_LEVEL=' + str(arguments.log_level))
```
Comment (elecro) [inline tools/scripts/build.py]: For me this looks better:
```python
opts = [
    '-DCMAKE_BUILD_TYPE=%s" % arguments.build_type.capitalize(),
    '-DUSE_%s=ON' % arguments.backend.upper(),
    '-DLOG_LEVEL=%d' % arguments.log_level
]
```
(of course the indent is aligned at the opening [.)

File: tools/scripts/build.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    if arguments.no_colored_logs:
+        opts.append('-DDISABLE_LOG_COLORS=ON')
+
+    if (arguments.backend != 'gles2'):
```
Comment (elecro) [inline tools/scripts/build.py]: No need for the parenthesis.

File: tools/scripts/unittest.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
+#! /usr/bin/python
+
+import argparse
+import subprocess
+import sys
+from os import path
+from os import chdir
```
Comment (elecro) [inline tools/scripts/unittest.py]: chdir is not used

**Comment by dbatyai** [comment]
Patch is updated

**Comment by szledan** [comment]
Great, thanks.

**Pull request merged**